src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,342 @@
+
+### Property access as a parameter
+
+For example, we can use the property access as a parameter.
+
+```javascript
+console.log(o2.foo);
+```
+
+In this case, the behavior is defined in the runtime semantics of `ArgumentList` production which calls `GetValue` on the argument:
+
+> [Runtime Semantics: ArgumentListEvaluation](https://tc39.es/ecma262/#sec-argument-lists-runtime-semantics-argumentlistevaluation)
+>
+> `ArgumentList : AssignmentExpression`
+>
+> 1. Let `ref` be the result of evaluating `AssignmentExpression`.
+> 2. Let `arg` be `? GetValue(ref)`.
+> 3. Return a List whose sole item is `arg`.
+
+`o2.foo` doesn't look like an `AssignmentExpression` but it is one, so this production is applicable. (Why `o2.foo` is an `AssignmentExpression` will be explained later in this post.)
+
+So, the `AssignmentExpression` is `o2.foo`. `ref`, the result of evaluating `o2.foo`, is the above mentioned Reference. Now we call `GetValue` on it.
+
+### Property access as the right hand side of an assignment
+
+We can also use the property access as a right hand side of an assignment:
+
+```javascript
+x = o2.foo;
+```
+
+In this case, the behavior is defined in the runtime semantics for the `AssignmentExpression : LeftHandSide = AssignmentExpression` production. Also this ends up calling `GetValue` on the result of evaluating the right hand side `AssignmentExpression`.
+
+> [Runtime Semantics: Evaluation for `AssignmentExpression : LeftHandSideExpression = AssignmentExpression`](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation)
+>
+> 1. If `LeftHandSideExpression` is neither an `ObjectLiteral` nor an `ArrayLiteral`, then
+>     1. Let `lref` be the result of evaluating `LeftHandSideExpression`.
+>     1. `ReturnIfAbrupt(lref)`.
+>     1. If `IsAnonymousFunctionDefinition(AssignmentExpression)` and `IsIdentifierRef` of `LeftHandSideExpression` are both `true`, then
+>         1. Let `rval` be `NamedEvaluation` of `AssignmentExpression` with argument `GetReferencedName(lref)`.
+>     1. Else,
+>         1. Let `rref` be the result of evaluating `AssignmentExpression`.
+>         1. Let `rval` be `? GetValue(rref)`.
+>     1. Perform `? PutValue(lref, rval)`.
+>     1. Return `rval`.
+> 1. Let `assignmentPattern` be the `AssignmentPattern` that is covered by `LeftHandSideExpression`.
+> 1. Let `rref` be the result of evaluating `AssignmentExpression`.
+> 1. Let `rval` be `? GetValue(rref)`.
+> 1. Perform `? DestructuringAssignmentEvaluation` of `assignmentPattern` using `rval` as the argument.
+> 1. Return `rval`.
+
+Now the `LeftHandSideExpression` (`x`) is not an object literal or an array literal, so we take the if branch in step 1. In step 1 a, we evaluate the left-hand side (`x`) and store the result into `lref`. `o2.foo` is not a function definition, so we don't take the if branch in 1 c, but the else branch in 1 d. There we evaluate `o2.foo` and call `GetValue` on it.
```

Comment:
This is a known issue - Mathias said he'll take care of the custom formatting. (The problem is making markdown do "1. foo a. subfoo i. subsubfoo" - style formatting. I have no idea how to, I tried but failed...)

Reply:
Looks like this will require either changing our CSS or using inline HTML like:

```html
<ol type="a">
<li>123
<li>456
</ol>
```

which is not as "good" as CSS version, but consistently renders as

<ol type="a">
<li>123
<li>456
</ol>

even on Github.

Reply:
(update based on the chat: found a better solution using `<style>` on the page to get consistent formatting; still not perfect, because, well, it uses `<style>` on the page, so looking into something more viable in `markdown-it`)

Reply:
@mathiasbynens On the second thought, do you think it would be useful to change styles by default on entire blog?

There already is (and probably will be) lots of quotes from the ECMAScript spec (duh, given the nature of the blog) and it could be nicer to keep nested list styles consistent everywhere.

FWIW this is what I did locally to style this one page and could add this to global CSS instead:
```css
ol > li > ol { list-style-type: lower-alpha; }
ol > li > ol > li > ol { list-style-type: lower-roman; }
```

Reply:
I was thinking an explicit wrapper a la `:::ecmascript-algorithm` (like `:::table-wrapper`) would be the best solution. We could then scope the styles to that wrapper, and we'd still be free to change the global styles. WDYT?

Reply:
PR: https://github.com/v8/v8.dev/pull/332

Reply:
We've merged #332; @marjakh can you please update the post using the mentioned wrapper?

Reply:
Done!